crates/ntx-builder/src/actor/account_state.rs

```diff
@@ -0,0 +1,348 @@
+use miden_node_proto::domain::note::{NetworkNote, SingleTargetNetworkNote};
+use miden_node_utils::tracing::OpenTelemetrySpanExt;
+use miden_objects::account::Account;
+use miden_objects::account::delta::AccountUpdateDetails;
+use miden_objects::block::{BlockHeader, BlockNumber};
+use miden_objects::note::{Note, Nullifier};
+use miden_objects::transaction::{PartialBlockchain, TransactionId};
+use tracing::instrument;
+
+use super::ActorShutdownReason;
+use super::note_state::{NetworkAccountNoteState, NetworkAccountUpdate};
+use crate::COMPONENT;
+use crate::actor::inflight_note::InflightNetworkNote;
+use crate::builder::ChainState;
+use crate::store::{StoreClient, StoreError};
+
+// TRANSACTION CANDIDATE
+// ================================================================================================
+
+/// A candidate network transaction.
+///
+/// Contains the data pertaining to a specific network account which can be used to build a network
+/// transaction.
+#[derive(Clone, Debug)]
+pub struct TransactionCandidate {
+    /// The current inflight state of the account.
+    pub account: Account,
+
+    /// A set of notes addressed to this network account.
+    pub notes: Vec<InflightNetworkNote>,
+
+    /// The latest locally committed block header.
+    ///
+    /// This should be used as the reference block during transaction execution.
+    pub chain_tip_header: BlockHeader,
+
+    /// The chain MMR, which lags behind the tip by one block.
+    pub chain_mmr: PartialBlockchain,
+}
+
+// NETWORK ACCOUNT STATE
+// ================================================================================================
+
+/// The current state of a network account.
+#[derive(Clone)]
+pub struct NetworkAccountState {
+    /// The network account prefix corresponding to the network account this state represents.
+    account_prefix: NetworkAccountPrefix,
+
+    /// Component of this state which Contains the committed and inflight account updates as well
+    /// as available and nullified notes.
+    account: NetworkAccountNoteState,
+
+    /// Uncommitted transactions which have some impact on the network state.
+    ///
+    /// This is tracked so we can commit or revert such transaction effects. Transactions _without_
+    /// an impact are ignored.
+    inflight_txs: BTreeMap<TransactionId, TransactionImpact>,
+
+    /// A set of nullifiers which have been registered for the network account.
+    nullifier_idx: HashSet<Nullifier>,
+}
+
+impl NetworkAccountState {
+    /// Maximum number of attempts to execute a network note.
+    const MAX_NOTE_ATTEMPTS: usize = 30;
+
+    /// Load's all available network notes from the store, along with the required account states.
+    #[instrument(target = COMPONENT, name = "ntx.state.load", skip_all)]
+    pub async fn load(
+        account: Account,
+        account_prefix: NetworkAccountPrefix,
+        store: &StoreClient,
+        block_num: BlockNumber,
+    ) -> Result<Self, StoreError> {
```

Comment:
Why do we need to pass `account_prefix` here as a parameter? We should be able to derive it from `account` - no? Passing both would create a potential for inconsistency when account and account prefix do not match.